crates/wasi/src/p2/pipe.rs

```diff
@@ -146,6 +146,68 @@ impl AsyncWrite for MemoryOutputPipe {
     }
 }
 
+pub trait SimpleCustomOutputWriter: Send + Sync + 'static {
```

Comment:
We've probably got 1-too-many traits already in the wasmtime-wasi crate, so mind moving this directly into the c-api crate? It should be fine to implement pollable/outputstream/asyncwrite directly on the c-api types and avoid the trait indirection here (should avoid arc/mutex as well I think)